Aplib.Core/Belief/SampledMemoryBelief.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using static Aplib.Core.Belief.UpdateMode;
+
+namespace Aplib.Core.Belief
+{
+    /// <summary>
+    /// The <see cref="SampledMemoryBelief{TReference, TObservation}"/> class represents the agent's belief of a single object,
+    /// but with additional "memory" of previous observations.
+    /// These observations are sampled at a fixed rate.
+    /// Some <i>object reference</i> is used to generate/update an <i>observation</i> 
+    /// (i.e., some piece of information on the game state as perceived by an agent).
+    /// This belief also stores a limited amount of previous observation samples in memory.
+    /// Optionally, the belief can always store the most recent observation, regardless of the sample rate.
+    /// </summary>
+    /// <remarks>
+    /// It supports implicit conversion to <typeparamref name="TObservation"/>.
+    /// </remarks>
+    /// <typeparam name="TReference">The type of the reference used to generate/update the observation.</typeparam>
+    /// <typeparam name="TObservation">The type of the observation the belief represents.</typeparam>
+    public class SampledMemoryBelief<TReference, TObservation> : MemoryBelief<TReference, TObservation>
+    {
+        /// <summary>
+        /// The sample interval of the memory (inverse of the sample rate).
+        /// One observation memory (i.e., snapshot) is stored every <see cref="_sampleInterval"/>-th cycle.
+        /// </summary>
+        private readonly int _sampleInterval;
+
+        /// <summary>
+        /// Specifies how this sampled memory belief should be updated.
+        /// </summary>
+        private readonly UpdateMode _updateMode;
+
+        private int _counter = 0;
+
+        /// <summary>
+        /// The number of cycles that have passed since the last memory sample was stored.
+        /// </summary>
+        private int Counter
+        {
+            get => _counter;
+            set => _counter = value % _sampleInterval;
```

Comment:
I understand where you are coming from. The idea of the count being cyclic (modulo) is there, such that ShouldSample() can test whether is has cycled back to 0. However, this mindset influences both this Counter and the ShouldSample method. Both are intertwined and depended on each other, hoping someone does not change one of both in isolation - whilst this might happen as its not that intuitive. Lastly, its a bit of code coupling.

In sake of seperation of concern, and my long story from just now, I would suggest making the counter quite litterally a counter (who knows, might be useful in the future) and change the ShouldSample to modulo himself. If your idea was to represent how many times it was since the most recent cycle, put that idea in a dedicated method or field.

TL;DR one strong idea is spread over two fields, where one might accidentally change since the idea and coupling is not that obvious, and break the idea.

Reply:
While I kinda agree with your point, I have decided to make it a cyclic (modulo) counter for the following reason. When the integer reaches the maximum value, it loops around and in the modulo class this might skip a few steps in the counter. It's very unlikely to happen, but I would rather do it like this. I will call it ModuloCounter tho.

Reply:
Oh you're right! Or BigInteger haha